group/views.py

```diff
@@ -104,39 +97,50 @@ def detail(request, group_id):
 
     group = get_group(group_id)
     show_number = 5; #number for brief list to show in group detail page.
-    all_contest = group.trace_contest.all()
+    all_contest = group.trace_contest.order_by('-start_time')
     annowence_list = group.announce.all()
-    student_list = group.member.order_by('user_level')
-    coowner_list = group.coowner.all()
+    student_list = group.member.order_by('username')
     owner = group.owner
-    user_is_owner = has_group_ownership(request.user, group)
+    form = AnnounceForm()
+
+    user = validate_user(request.user)
+    user_is_owner = has_group_ownership(user, group)
+    user_is_coowner = has_group_coownership(user, group)
+
+    user_has_auth = user_is_owner or user_is_coowner
 
     running_contest_list = []
     ended_contest_list = []
     now = timezone.now()
-    for contest in all_contest:
-        if contest.start_time < now and contest.end_time > now:
-            running_contest_list.append(contest)
-        elif contest.end_time < now:
-            ended_contest_list.append(contest)
+    running_contest_list = group.trace_contest.filter(start_time__lte=now, end_time__gte=now)[0:show_number]
+    ended_contest_list = group.trace_contest.filter(end_time__lte=now)[0:show_number]
+
+    student_list = get_current_page(request, student_list)
 
     return render_index(
         request, 'group/groupDetail.html', {
-            'rc_list': running_contest_list[0:show_number],
-            'ec_list': ended_contest_list[0:show_number],
-            'an_list': annowence_list,
-            'coowner_list': coowner_list,
+            'running_contest_list': running_contest_list,
+            'ended_contest_list': ended_contest_list,
+            'announce_list': annowence_list,
             'owner': owner,
-            's_list': student_list,
-            'group_name': group.gname,
-            'group_description': group.description,
-            'group_id': group.id,
-            'user_is_owner': user_is_owner,
+            'student_list': student_list,
+            'group': group,
+            'user_has_auth': user_has_auth,
+            'form': form,
+            'redirect_page' : 'detail',
         })
 
-
 def list(request):
     all_group = Group.objects.order_by('id')
+    all_group = get_current_page(request, all_group)
+
+    return render_index(
+        request,'group/groupList.html', {
+            'all_group_list': all_group,
+            'include_flag': 'all_group',
+        })
+
+def my_list(request):
     if request.user.is_anonymous():
```

Comment:
Since you hide my_group tab when user not login, this if statement is not needed anymore.
BTW, I prefer to use ```@login_required()``` for this function to avoid unexpected error.